libnetwork/drivers/bridge/setup_ip_tables_linux_test.go

```diff
@@ -184,3 +204,150 @@ func TestSetupIP6TablesWithHostIPv4(t *testing.T) {
+			wantIPv4Snat:       true,
+		},
+		{
+			desc:               "IPv4 masquerade, IPv6 masquerade",
+			enableIPTables:     true,
+			enableIP6Tables:    true,
+			enableIPv6:         true,
+			enableIPMasquerade: true,
+			wantIPv4Masq:       true,
+			wantIPv6Masq:       true,
+		},
+		{
+			desc:               "IPv4 SNAT, IPv6 masquerade",
+			enableIPTables:     true,
+			enableIP6Tables:    true,
+			enableIPv6:         true,
+			enableIPMasquerade: true,
+			hostIPv4:           hostIPv4,
+			wantIPv4Snat:       true,
+			wantIPv6Masq:       true,
+		},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			defer netnsutils.SetupTestOSContext(t)()
+			dc := &configuration{
+				EnableIPTables:  tc.enableIPTables,
+				EnableIP6Tables: tc.enableIP6Tables,
+			}
+			r := &testRegisterer{t: t}
+			if err := Register(r, map[string]interface{}{netlabel.GenericData: dc}); err != nil {
+				t.Fatal(err)
+			}
+			if r.d == nil {
+				t.Fatal("testRegisterer.RegisterDriver never called")
+			}
+			nc := &networkConfiguration{
+				BridgeName:         br,
+				AddressIPv4:        brIPv4,
+				AddressIPv6:        brIPv6,
+				EnableIPv6:         tc.enableIPv6,
+				EnableIPMasquerade: tc.enableIPMasquerade,
+				HostIPv4:           tc.hostIPv4,
+			}
+			ipv4Data := []driverapi.IPAMData{{Pool: maskedBrIPv4, Gateway: brIPv4}}
+			ipv6Data := []driverapi.IPAMData{{Pool: maskedBrIPv6, Gateway: brIPv6}}
+			if !nc.EnableIPv6 {
+				nc.AddressIPv6 = nil
+				ipv6Data = nil
+			}
+			if err := r.d.CreateNetwork("nattest", map[string]interface{}{netlabel.GenericData: nc}, nil, ipv4Data, ipv6Data); err != nil {
+				t.Fatal(err)
+			}
+			defer func() {
+				if err := r.d.DeleteNetwork("nattest"); err != nil {
+					t.Fatal(err)
+				}
+			}()
+			// Log the contents of all chains to aid troubleshooting.
+			for _, ipv := range []iptables.IPVersion{iptables.IPv4, iptables.IPv6} {
+				ipt := iptables.GetIptable(ipv)
+				for _, table := range []iptables.Table{iptables.Nat, iptables.Filter, iptables.Mangle} {
+					out, err := ipt.Raw("-t", string(table), "-S")
+					if err != nil {
+						t.Error(err)
+					}
+					t.Logf("%s: %s %s table rules:\n%s", tc.desc, ipv, table, string(out))
+				}
+			}
+			for _, rc := range []struct {
+				want bool
+				rule iptRule
+			}{
+				// Order doesn't matter: At most one of the following IPv4 rules will exist.  Same goes for
+				// the IPv6 rules.
+				{tc.wantIPv4Masq, iptRule{iptables.IPv4, iptables.Nat, "POSTROUTING", []string{"-s", maskedBrIPv4.String(), "!", "-o", br, "-j", "MASQUERADE"}}},
+				{tc.wantIPv4Snat, iptRule{iptables.IPv4, iptables.Nat, "POSTROUTING", []string{"-s", maskedBrIPv4.String(), "!", "-o", br, "-j", "SNAT", "--to-source", hostIPv4.String()}}},
+				{tc.wantIPv6Masq, iptRule{iptables.IPv6, iptables.Nat, "POSTROUTING", []string{"-s", maskedBrIPv6.String(), "!", "-o", br, "-j", "MASQUERADE"}}},
+			} {
+				// No state is changed so there's no need to call netnsutils.SetupTestOSContext.
```

Comment:
State is change on line 318, which is fine because netnsutils.SetupTestOSContext is called on line 292. The only reason I brought it to your attention earlier was to let you know that you could safely assert on the contents of iptables chains verbatim.
```suggestion
```

Reply:
Ah, gotcha.  Done.